### PR TITLE
Adding modifiable state in Dependency object. Use to capture symlink …

### PR DESCRIPTION
**Pull request by mortenvp**
…information

### CODE REVIEW

File: src/wurf/registry.py [hunk 1]
```diff
@@ -564,6 +573,9 @@ def active_dependency_resolver(registry, dependency):
     resolver = CreateSymlinkResolver(
         resolver=resolver, dependency=dependency, symlinks_path=symlinks_path,
         ctx=ctx)
+        
+    resolver = ContextMsgResolver(resolver=resolver, ctx=ctx,
```
Comment (petya2164) [inline src/wurf/registry.py]: This is an excellent place for the ContextMsgResolver, great simplification!

File: src/wurf/dependency.py [hunk 2]
```diff
@@ -30,16 +31,26 @@ def __init__(self, **kwargs):
...
 
+        Additional attributes may be added to the dependency during resolve. 
+        However, they cannot overwrite the existing attributes created at 
+        construction time.
+        
+        So in the small example from before all the following attributes are 
+        read-only: name, recurse, optional, resolver, method, checkout, 
+        sources and sha1.
+        
+        Any other attributes can be added an modified as needed.
```
Comment (petya2164) [inline src/wurf/dependency.py]: added **and** modified

File: src/wurf/dependency.py [hunk 3]
```diff
@@ -79,34 +91,52 @@ def __init__(self, **kwargs):
         #    in external formats such as files
         # 2. Use the Python hash() function to generate a integer hash for 
         #    run-time usage (i.e. within the same interpreter session).
-        self.hash = None
+        #super(Dependency, self).__setattr__('hash', None)
```
Comment (petya2164) [inline src/wurf/dependency.py]: I guess we can remove this line.

File: src/wurf/dependency.py [hunk 4]
```diff
@@ -79,34 +91,52 @@ def __init__(self, **kwargs):
...
+    def __setattr__(self, name, value): 
+        """ Sets a resolver attribute."""
 
-        if option in self.info:
-            return self.info[option]
+        if name in self.read_only:
+            raise AttributeError("Attribute {} read-only.".format(name))
         else:
-            raise AttributeError("No such attribute: " + option)
+            self.read_write[name] = value
```
Comment (petya2164) [inline src/wurf/dependency.py]: Do we have end-of-line whitespace here?

File: src/wurf/dependency.py [hunk 5]
```diff
@@ -79,34 +91,52 @@ def __init__(self, **kwargs):
...
         #    run-time usage (i.e. within the same interpreter session).
-        self.hash = None
+        #super(Dependency, self).__setattr__('hash', None)
+        self.read_only['hash'] = None
+        
+        # Resolver attributes (modifiable)
+        object.__setattr__(self, 'read_write', dict())
+        
         
     def __getattr__(self, option):
```
Comment (petya2164) [inline src/wurf/dependency.py]: I don't really think that `option` is a good name here. I would go for `property` or `attribute` instead.

File: src/wurf/dependency.py [hunk 6]
```diff
@@ -79,34 +91,52 @@ def __init__(self, **kwargs):
...
+            return self.read_only[option]
+
+        elif option in self.read_write:
+            return self.read_write[option]
+        
+        else:
+            return None
+
+
+    def __setattr__(self, name, value):
```
Comment (petya2164) [inline src/wurf/dependency.py]: Here you used `name` instead of `option`. Again, `property` would feel more accurate.

File: src/wurf/dependency.py [hunk 7]
```diff
@@ -79,34 +91,52 @@ def __init__(self, **kwargs):
...
-            raise AttributeError("No such attribute: " + option)
+            self.read_write[name] = value   
 
     def __contains__(self, option):
         """ Checks if the option is available.
         
         :return: True if the option is available otherwise False
         """
-        return option in self.info
+        return (option in self.read_only) or (option in self.read_write)
```
Comment (petya2164) [inline src/wurf/dependency.py]: option => property

File: src/wurf/context_msg_resolver.py [hunk 8]
```diff
@@ -26,9 +24,10 @@ def resolve(self):
 
         :return: The path as a string.
         """
+        
 
         self.ctx.start_msg('{} {}'.format(
-            self.method, self.dependency.name))
+            self.dependency.resolver_method, self.dependency.name))
```
Comment (petya2164) [inline src/wurf/context_msg_resolver.py]: I would like to propose an alternative start_msg format, just to make things understandable for the user. The current format looks confusing if the resolver_method comes first.
If it is an **active** step where we actually resolve something:
`Resolve "waf-tools" (git semver/git checkout/user path)`
I think we can keep Load for the passive step, but I am not fully happy with Load (help) :
`Load "waf-tools"` or `Load (help) "waf-tools"`

Of course, this would require that we define a separate `resolver_action` (Resolve/Load) and a `resolver_method` which is only needed for an active step.

File: src/wurf/dependency.py [hunk 9]
```diff
@@ -79,34 +91,52 @@ def __init__(self, **kwargs):
...
+        return "Dependency\ninfo: {}\nresolver info: {}".format(
+            self.read_only, self.read_write)
 
     def __hash__(self):
         """ :return: Integer hash value for the dependency. """
         
-        if not self.hash:
-            self.hash = hash(self.info['sha1'])
+        if not self.read_only['hash']:
+            self.read_only['hash'] = hash(self.read_only['sha1'])
```
Comment (petya2164) [inline src/wurf/dependency.py]: I am just curious: Do we use this hash value for anything besides comparing 2 dependencies in dependency_manager?:
`if seen_dependency.sha1 != dependency.sha1:`
They are not stored in any dict with their hash value, right?

If the comparison is the only use case, then I think that the hash might be too strict down the line. Just consider all the recursion that happens in our projects. If there is a slight mismatch, the comparison fails. 
For example, let's say that I add a new source repo to a given dependency, because we moved it to a new location. Then I have to make the same modification in all the wscripts where it is used, otherwise I get a mismatch.

This is beyond the scope of this PR, but I think we need to consider a smarter way of determining compatibility.

File: src/wurf/dependency.py (hunk 2 above)
Comment (mortenvp) [inline src/wurf/dependency.py]: :+1:

File: src/wurf/registry.py (hunk 1 above)
Comment (mortenvp) [inline src/wurf/registry.py]: :+1:

File: src/wurf/dependency.py (hunk 3 above)
Comment (mortenvp) [inline src/wurf/dependency.py]: :+1:

File: src/wurf/dependency.py (hunk 4 above)
Comment (mortenvp) [inline src/wurf/dependency.py]: :+1:

File: src/wurf/dependency.py (hunk 5 above)
Comment (mortenvp) [inline src/wurf/dependency.py]: Yes, the names in this class are not really good.. I will go with `attribute` since `property` is also a python built-in

File: src/wurf/dependency.py (hunk 6 above)
Comment (mortenvp) [inline src/wurf/dependency.py]: :+1:

File: src/wurf/dependency.py (hunk 7 above)
Comment (mortenvp) [inline src/wurf/dependency.py]: :+1:

File: src/wurf/context_msg_resolver.py (hunk 8 above)
Comment (mortenvp) [inline src/wurf/context_msg_resolver.py]: Alright, I think it would be good to find a nicer way to print this information. Should we also do a separate PR for that?

File: src/wurf/dependency.py (hunk 9 above)
Comment (mortenvp) [inline src/wurf/dependency.py]: The hash is also used by the registry caching functionality. 

In principle we can freely decide how what information should be used when calculating the hash. However, I would rather be very strict now.. see if it becomes a problem and the fix it. I worry that if it is not strict small difficult bugs could creep in where you do not see a mismatch between two dependencies causing a problem.. However, disregarding for example the `sources` attribute could potentially be safe.

File: src/wurf/context_msg_resolver.py (hunk 8 above)
Comment (petya2164) [inline src/wurf/context_msg_resolver.py]: Yes, this is beyond the scope of this branch. I will experiment with this in another branch.

File: src/wurf/dependency.py (hunk 9 above)
Comment (petya2164) [inline src/wurf/dependency.py]: I agree about potentially disregarding the sources, but let's move this discussion to #waf.

**Comment by petya2164** [comment]
We can merge this one!

**Comment by mortenvp** [comment]
Will do! :+1:
